File: src/eu/yeger/gramofo/fol/formula/FOLQuantifier.java
package eu.yeger.gramofo.fol.formula;

import java.util.Arrays;
import java.util.LinkedHashSet;

public class FOLQuantifier extends FOLFormula {

    FOLQuantifier(FOLType type, FOLBoundVariable variable, FOLFormula operand, boolean hasBrackets, boolean hasDot, String name) {
        super(type, new LinkedHashSet<>(Arrays.asList(variable, operand)), hasBrackets, hasDot, name);
    }

    ///TODO
    @Override
    public String getFormulaString() {

        StringBuilder sb = new StringBuilder();
        sb.append(getName());
        sb.append(getChildAt(0));

        return sb.toString();
    }

    @Override
    public String getFormulaStringForDepth(int currentDepth, int maxDepth) {
        if (currentDepth >= maxDepth && maxDepth != -1) {
            return DOT;
        }
        FOLFormula child0 = getChildAt(0);
        FOLFormula child1 = getChildAt(1);
        StringBuilder sb = new StringBuilder();
        sb.append(getName());
        sb.append(child0);
        if (!child1.getHasDot() && !isUnary(child1)) {
            sb.append(" ");
        }
        sb.append(child1.getFormulaStringForDepth(currentDepth, maxDepth));
        maybeWrapBracketsAndDot(sb);
        return sb.toString();
    }

    @Override
    public String toString() {
        FOLFormula child0 = getChildAt(0);
        FOLFormula child1 = getChildAt(1);
        StringBuilder sb = new StringBuilder();
        sb.append(getName());
        sb.append(child0);
        if (!child1.getHasDot() && !isUnary(child1)) {
            sb.append(" ");
        }
        sb.append(child1);
        maybeWrapBracketsAndDot(sb);
        return sb.toString();
    }

    public boolean isUnary(FOLFormula formula) {
        FOLType type = formula.getType();
        return Arrays.asList(FOLType.Exists, FOLType.Forall, FOLType.Not).stream().anyMatch(forbiddenType -> forbiddenType == type);
    }

}

File: src/eu/yeger/gramofo/fol/formula/FOLQuantifier.kt
package eu.yeger.gramofo.fol.formula

import eu.yeger.gramofo.fol.ModelCheckerTrace
import eu.yeger.gramofo.fol.SymbolTable
import eu.yeger.gramofo.model.domain.Graph
import eu.yeger.gramofo.model.domain.Node
import eu.yeger.gramofo.model.dto.TranslationDTO

sealed class FOLQuantifier(
    name: String,
    val variable: FOLBoundVariable,
    val operand: FOLFormula,
) : FOLFormula(name) {

    class Existential(variable: FOLBoundVariable, operand: FOLFormula) : FOLQuantifier(EXISTS, variable, operand) {
        override fun checkModel(
            graph: Graph,
            symbolTable: SymbolTable,
            variableAssignments: Map<String, Node>,
            shouldBeModel: Boolean,
        ): ModelCheckerTrace {
            val children = graph.nodes.map { node: Node ->
                operand.checkModel(graph, symbolTable, variableAssignments + (variable.name to node), shouldBeModel)
            }
            return if (children.any(ModelCheckerTrace::isModel)) {
                validated(TranslationDTO("api.exists.valid"), variableAssignments, shouldBeModel, *children.toTypedArray())
            } else {
                invalidated(TranslationDTO("api.exists.invalid"), variableAssignments, shouldBeModel, *children.toTypedArray())
            }
        }
    }

    class Universal(variable: FOLBoundVariable, operand: FOLFormula) : FOLQuantifier(FOR_ALL, variable, operand) {
        override fun checkModel(
            graph: Graph,
            symbolTable: SymbolTable,
            variableAssignments: Map<String, Node>,
            shouldBeModel: Boolean,
        ): ModelCheckerTrace {
            val children = graph.nodes.map { node: Node ->
                operand.checkModel(graph, symbolTable, variableAssignments + (variable.name to node), shouldBeModel)
            }
            return if (children.all(ModelCheckerTrace::isModel)) {
                validated(TranslationDTO("api.forall.valid"), variableAssignments, shouldBeModel, *children.toTypedArray())
            } else {
                invalidated(TranslationDTO("api.forall.invalid"), variableAssignments, shouldBeModel, *children.toTypedArray())
            }
        }
    }

    override fun getFormulaString(variableAssignments: Map<String, Node>): String {
        return buildString {
            append(name)
            append(variable.getFormulaString(variableAssignments))
            if (!operand.hasDot && !isUnary(operand)) {
                append(" ")
            }
            append(operand.getFormulaString(variableAssignments))
            maybeWrapBracketsAndDot()
        }
    }

    private fun isUnary(formula: FOLFormula): Boolean {
        return when (formula) {
            is Existential -> true
            is Universal -> true
            is FOLOperator.Unary.Not -> true
            else -> true
        }
    }
}
